File: php/services/SmartHomeApp.php
<?php

namespace services;

use controllers\LightController;
use controllers\ThermostatController;
use controllers\SecurityController;
use devices\IDevice;

class SmartHomeApp
{
    public function __construct(
        private LightController $lightController,
        private ThermostatController $thermostatController,
        private SecurityController $securityController,
        private array $devices
    ) {}

    public function turnOnLight(string $room): void
    {
        $this->lightController->turnOnLight($room);
    }

    public function setTemperature(float $temp): void
    {
        $this->thermostatController->setTemperature($temp);
    }

    public function lockDoors(): void
    {
        $this->securityController->lockDoors();
    }

    public function turnOffAll(): void
    {
        echo "Extinction globale..." . PHP_EOL;

        foreach ($this->devices as $device) {
            if ($device instanceof IDevice) {
                $device->turnOff();
            }
        }
    }
}

File: java/src/services/SmartHomeApp.java
package services;

import java.util.List;

import controllers.LightController;
import controllers.SecurityController;
import controllers.ThermostatController;
import devices.IDevice;

public class SmartHomeApp {
    private LightController lightController;
    private ThermostatController thermostatController;
    private SecurityController securityController;
    private List<IDevice> devices;

    public SmartHomeApp(
            LightController lightController,
            ThermostatController thermostatController,
            SecurityController securityController,
            List<IDevice> devices
    ) {
        this.lightController = lightController;
        this.thermostatController = thermostatController;
        this.securityController = securityController;
        this.devices = devices;
    }

    public void turnOnLight(String room) {
        lightController.turnOnLight(room);
    }

    public void setTemperature(float temp) {
        thermostatController.setTemperature(temp);
    }

    public void lockDoors() {
        securityController.lockDoors();
    }

    public void turnOffAll() {
        System.out.println("Extinction globale...");

        for (IDevice device : devices) {
            device.turnOff();
        }
    }
}